### DISCUSSION TITLE
Homebrew and universal2/fat binaries

### DISCUSSION THREAD

**Discussion by datalowe** [Homebrew/discussions#3706] [Everyday usage] [answered]
Hi,

I'd like to start off saying that this is not a feature request or issue report, only a description of something that can be confusing and a wish to discuss how Homebrew currently handles it and might handle it in the future.

As far as I understand, on MacOS, Homebrew sometimes (always?) installs architecture-specific libraries when running brew install. As an example: `brew install openssl@1.1`. This becomes apparent when checking the installed libraries with `lipo` or `file`. Continuing the example, `lipo -detailed_info /opt/homebrew/opt/openssl@1.1/lib/libssl.a` produces this output:
```sh
input file /opt/homebrew/opt/openssl@1.1/lib/libssl.a is not a fat file
Non-fat file: /opt/homebrew/opt/openssl@1.1/lib/libssl.a is architecture: arm64
```
The output indicates that the binaries are only compatible with the arm64 (M1) architecture. This is unlike so-called 'universal2' or 'fat' binaries, which work with _both_ arm64 and x86_64 (Intel) architectures (at the cost of larger filesizes due to requiring 'duplicated' code under the hood, IIUC). For most users, this isn't a problem. However, it does turn into an issue if one wishes to rely/depend on brew-installed packages when building other universal2 binaries. The particular issue I ran into was installing a universal2 Python version with pyenv (this is also reported in a [pyenv GH issue comment](https://github.com/pyenv/pyenv/issues/1877#issuecomment-1088127640)) when attempting to use OpenSSL libraries installed with homebrew. It took me a long time to understand what was going on, and as far as I can understand, the only way around this issue at the moment is to download the OpenSSL source and compile universal2 binaries myself.

Here are a few questions I would be happy to hear thoughts on, if possible. This would be especially relevant for better understanding how one should or shouldn't use Homebrew as part of build toolchains.
* Does Homebrew provide any flags or the like for choosing to install universal2 binaries, or are there any plans for this in the future?
* Does Homebrew raises this tricky question somewhere in the documentation, and/or give instructions how best to go about dealing with this issue?
* Is Homebrew something that one should/shouldn't plan on using in build chains that are to result in universal2 binaries?

**Comment by SMillerDev** [comment, marked as answer]
> Does Homebrew provide any flags or the like for choosing to install universal2 binaries, or are there any plans for this in the future?

No, because homebrew is a package manager not a dependency manager.

> Does Homebrew raises this tricky question somewhere in the documentation, and/or give instructions how best to go about dealing with this issue?

No, for the same reason the kernel docs don't explain how to make coffee. It's not a goal of the project.
> Is Homebrew something that one should/shouldn't plan on using in build chains that are to result in universal2 binaries?

Since there aren't any plans to support this, ever I'd recommend against it.

**Reply by datalowe** [reply]
Great, thanks for the answers :) For anyone stumbling upon this thread, a relevant discussion regarding asking pyenv to not rely on Homebrew (to avoid issues similar to what I described in OP) can be found here: https://github.com/pyenv/pyenv/pull/2464
